src/relations/postgresql_provider.py

```diff
@@ -0,0 +1,154 @@
+            relation_name: the name of the relation
+        """
+        self.relation_name = relation_name
+
+        super().__init__(charm, self.relation_name)
+        self.framework.observe(
+            charm.on[self.relation_name].relation_broken, self._on_relation_broken
+        )
+
+        self.charm = charm
+
+        # Charm events defined in the database provides charm library.
+        self.database_provides = DatabaseProvides(self.charm, relation_name=self.relation_name)
+        self.framework.observe(
+            self.database_provides.on.database_requested, self._on_database_requested
+        )
+
+    def _on_database_requested(self, event: DatabaseRequestedEvent) -> None:
+        """Generate password and handle user and database creation for the related application."""
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
+            return
+
+        # Retrieve the database name and extra user roles using the charm library.
+        database = event.database
+        extra_user_roles = event.extra_user_roles
+
+        try:
+            # Creates the user and the database for this specific relation.
+            user = f"relation_id_{event.relation.id}"
+            password = new_password()
+            self.charm.postgresql.create_user(user, password, extra_user_roles=extra_user_roles)
+            self.charm.postgresql.create_database(database, user)
+
+            # Share the credentials with the application.
+            self.database_provides.set_credentials(event.relation.id, user, password)
+
+            # Update the read/write and read-only endpoints.
+            self.update_endpoints(event)
+
+            # Set the database version.
+            self.database_provides.set_version(
+                event.relation.id, self.charm.postgresql.get_postgresql_version()
+            )
+        except (
+            PostgreSQLCreateDatabaseError,
+            PostgreSQLCreateUserError,
+            PostgreSQLGetPostgreSQLVersionError,
+        ) as e:
+            logger.exception(e)
+            self.charm.unit.status = BlockedStatus(
+                f"Failed to initialize {self.relation_name} relation"
+            )
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Remove the user created for this relation."""
+        # Check for some conditions before trying to access the PostgreSQL instance.
+        if (
+            "cluster_initialised" not in self.charm._peers.data[self.charm.app]
+            or not self.charm._patroni.member_started
+        ):
+            event.defer()
+            return
+
+        if not self.charm.unit.is_leader():
```

Comment:
Same comment as above about leader test.

Reply:
Fixed in https://github.com/canonical/postgresql-operator/pull/15/commits/d1a0911f6ce90104c34b55d2eaea4f4e68bcb3ea,